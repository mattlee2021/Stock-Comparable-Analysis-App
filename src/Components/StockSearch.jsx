import { useState } from "react";
import styles from "./StockSearch.module.css";
import NewTableButton from "./NewTableButton";
import stockApiService from "./stockApi.service";

const StockSearch = (props) => {
  const [showSimilarStocks, setShowSimilarStocks] = useState(true);
  const [suggestedResults, setSuggestedResults] = useState([]);
  const [ticker, setTicker] = useState("");
  let searchResults;

  const addStockToList = async () => {
    const stockToAdd = await stockApiService.getStockData(ticker);
    if (stockToAdd) {
      props.getStockData(stockToAdd);
    }
  };

  const addSimilarStockToList = async () => {
    const similarStockData = await stockApiService.getSimilarStocks(ticker);
    if (similarStockData) {
      similarStockData.forEach((stockData) => {
        props.getStockData(stockData);
      });
    }
  };

  const handleSubmitTicker = async (event) => {
    event.preventDefault();
    setTicker("");
    setSuggestedResults(() => []);
    addStockToList();
    if (showSimilarStocks) {
      addSimilarStockToList();
    }
  };

  const onChangeInput = async (event) => {
    const input = event.target.value;
    setTicker(() => input);
    setSuggestedResults(() => []);
    searchResults = await stockApiService.getMatchingStockNames(input);
    if (searchResults) {
      setSuggestedResults(() => searchResults);
    }
  };

  return (
    <form>
      <div className={styles.flexContainer}>
        <div className={styles.searchContainer}>
          <div className={styles.inputContainer}>
            <label className={styles.tickerLabel}>Ticker</label>
            <input
              type="text"
              value={ticker}
              onChange={onChangeInput}
              className={styles.searchBar}
            />
            <button
              type="submit"
              disabled={ticker.length === 0}
              onClick={handleSubmitTicker}
              className={
                ticker.length === 0
                  ? styles.tickerButtonDisabled
                  : styles.tickerButton
              }
            >
              Submit
            </button>
          </div>
          <div className={styles.searchResults}>
            {suggestedResults.map((stock) => {
              return (
                <ul
                  onClick={() => {
                    setTicker(() => stock[0]);
                    setSuggestedResults(() => []);
                  }}
                >
                  Symbol: {stock[0]} Name: {stock[1]}
                </ul>
              );
            })}
          </div>
        </div>

        <div className={styles.featureContainer}>
          <div className={styles.similarStockFlag}>
            Show Similar Stocks
            <input
              type="checkbox"
              checked={showSimilarStocks}
              onChange={() => {
                setShowSimilarStocks((prev) => {
                  return !prev;
                });
              }}
            />
          </div>
          <NewTableButton
            setTableData={props.setTableData}
            setSelectedTableIndex={props.setSelectedTableIndex}
          />
        </div>
      </div>
    </form>
  );
};

export default StockSearch;
