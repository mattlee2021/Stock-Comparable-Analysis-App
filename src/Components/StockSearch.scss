form {
  position: relative;
  font-family: Arial, Helvetica, sans-serif;
  min-height: 185px;
}

.ticker {
  position: relative;
  padding-top: 10px;
  display: inline-block;
  width: 50%;
  margin: 0 30%;
  //margin: 0 30%;
}

.ticker button {
  border-radius: 8px;
  margin: auto;
  border: 3px solid black;
  color: black;
  position: absolute;
  background-color: white;
  &.disabled {
    border: 3px solid #999999;
    background-color: #cccccc;
    color: #666666;
  }
  cursor: pointer;
}

.ticker label {
  font-size: 20px;
}

.similarStock {
  display: inline-block;
  margin-left: 5rem;
}

.similarStock label {
  font-size: 17px;
}

.similarStock input {
  cursor: pointer;
}

.searchResults {
  font-size: 14px;
  width: 50%;
  left: 32%;
  
  position: relative;
  //right: 20px;
  cursor: pointer;
}

.searchBar {
  border-radius: 4px;
  font-size: 15px;
  margin: 0 1rem;
  width: 15rem;
}


