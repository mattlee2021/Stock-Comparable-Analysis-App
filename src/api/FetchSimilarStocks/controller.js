import SimilarStocks from "../services/similarTickers.service.js";

class Controller {
  async getSimilarTickers(req, res) {
    try {
      const { ticker } = req.body;
      const response = await SimilarStocks.getSimilarStocks(ticker);
      res.status(200).send(response);
    } catch (error) {
      const errorResponse = {
        status: "fail",
        data: { "error message": "API Error" },
      };
      res.status(404).send(errorResponse);
    }
  }
}

export default new Controller();
