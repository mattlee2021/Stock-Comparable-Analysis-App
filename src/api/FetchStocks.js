const FetchStocks = (ticker, applyData, similarStockRequest) => {
  fetch(
    "https://www.alphavantage.co/query?function=OVERVIEW&symbol=" +
      ticker +
      "&apikey=ZY9GZNYZQM8C1MQC"
  ) //Amherst Key ZY9 GZN YZQ M8C 1MQC
    //TD8ZNN64UTNOK6DA
    .then((response) => {
      console.log("Got a response,", response);
      if (response) {
        console.log("Sending a response");
        return response.json();
      } else {
        console.log("Reached rate limit");
        throw new Error("Error here!!");
      }
    })
    .then((data) => {
      console.log("New Data,", data);
      if (
        data.Name === undefined &&
        data.Note !== undefined &&
        !similarStockRequest
      ) {
        throw new Error(
          "You hit the rate limit of 5 stock requests per minute. Please wait another minute before adding new stocks."
        );
      } else if (data.Name === undefined && similarStockRequest) {
        return;
      } else if (data.Name === undefined) {
        throw new Error("Invalid Ticker");
      }

      applyData(
        {
          Ticker: data.Symbol,
          Name: data.Name,
          "P/E": data.PERatio,
          "P/B": data.PriceToBookRatio,
          EPS: data.EPS,
          "EV/EBITDA": data.EVToEBITDA,
          "EV/Revenue": data.EVToRevenue,
          "Profit Margin": data.ProfitMargin,
          Sector: data.Sector,
        },
        similarStockRequest
      );
    })
    .catch((error) => {
      alert(error);
    });
};

export default FetchStocks;
