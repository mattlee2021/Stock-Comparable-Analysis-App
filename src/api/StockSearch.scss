// form {
//   display: flex;
//   flex-direction: column;
//   justify-content: center;
//   align-content: flex-start;
//   width: 10%;
//   text-align: center;
//   margin: 1em auto 1em auto;
// }

// label {
//   font-size: 20px;
//   font-family: Arial, Helvetica, sans-serif;
// }

form {
  position: relative;
  font-family: Arial, Helvetica, sans-serif;
}

.ticker {
  position: absolute;
  left: 40%;
  padding-top: 10px;
  display: flex;
  gap: 2%;
}

.ticker button {
  width: 25%;
  border-radius: 8px;
  margin: auto;
  border: 3px solid black;
  color: black;
  background-color: white;
  &.disabled {
    border: 3px solid #999999;
    background-color: #cccccc;
    color: #666666;
  }
}

.ticker label {
  font-size: 20px;
}

.ticker input {
  border-radius: 4px;
  font-size: 15px;
  width: 50%;
}

.similarStock {
  position: absolute;
  left: 44%;
  top: 40px;
}

.similarStock label {
  font-size: 17px;
}

.similarStock input {
}
