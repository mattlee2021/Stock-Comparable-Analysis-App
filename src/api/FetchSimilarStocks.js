import FetchStocks from "./FetchStocks";

const FetchSimilarStocks = (ticker, applyData) => {
  fetch(
    "https://stock-data-yahoo-finance-alternative.p.rapidapi.com/v6/finance/recommendationsbysymbol/" +
      ticker,
    {
      method: "GET",
      headers: {
        "x-rapidapi-host":
          "stock-data-yahoo-finance-alternative.p.rapidapi.com",
        "x-rapidapi-key": "80ad986cb5mshcabba16d2878f8fp14dca4jsn61fd4cc10485",
      },
    }
  )
    .then((response) => {
      if (response.ok) {
        return response.json();
      } else {
        throw new Error("No similar companies can be obtained for this stock.");
      }
    })
    .then((data) => {
      console.log("Retrieved Data", data.finance.result[0].recommendedSymbols);
      const similarStocks = data.finance.result[0].recommendedSymbols;
      /*
       data.finance.result[0].recommendedSymbols returns an array of 5 similar stocks in order from most similar (index 0) to less similar (index 4),
       so we are limiting the number of similar stocks by 1 to have a total of 4 stocks sent to FetchStocks(). 
       Therefore, for every similar stock request we will have the original stock request and 4 similar stock requests 
       for a total of 5 stock requests for FetchStocks. There is a max of 5 stock requests/minute. 
       */
      for (let i = 0; i < similarStocks.length - 1; i++) {
        FetchStocks(similarStocks[i].symbol, applyData, true);
      }
    })
    .catch((error) => {
      alert(error);
    });
};

export default FetchSimilarStocks;
