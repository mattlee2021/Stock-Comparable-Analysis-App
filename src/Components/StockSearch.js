import { useState } from "react";
import "./StockSearch.scss";
import FetchStocks from "../api/FetchStocks";
import FetchSimilarStocks from "../api/FetchSimilarStocks";
import FetchStockNames from "../api/FetchStockNames";

const StockSearch = (props) => {
  const [showSimilarStocks, setShowSimilarStocks] = useState(true);
  const [suggestedResults, setSuggestedResults] = useState([]);
  const [ticker, setTicker] = useState("");
  let searchResults;

  //API KEY is TD8ZNN64UTNOK6DA

  const addStockToList = async () => {
    const stockToAdd = await FetchStocks(ticker);
    if (stockToAdd) {
      props.getStockData(stockToAdd);
    }
  };

  const addSimilarStockToList = async () => {
    const similarStockData = await FetchSimilarStocks(ticker);
    if (similarStockData) {
      similarStockData.forEach((stockData) => {
        props.getStockData(stockData);
      });
    }
  };

  const handleSubmitTicker = async (event) => {
    event.preventDefault();
    setTicker("");
    setSuggestedResults(() => []);
    addStockToList();
    if (showSimilarStocks) {
      addSimilarStockToList();
    }
  };

  const onChangeInput = async (event) => {
    const input = event.target.value;
    setTicker(() => input);
    setSuggestedResults(() => []);
    searchResults = await FetchStockNames(input);
    if (searchResults) {
      setSuggestedResults(() => searchResults);
    }
  };

  return (
    <form>
      <div className="ticker">
        <label>Ticker</label>
        <input
          type="text"
          value={ticker}
          onChange={onChangeInput}
          className="searchBar"
        />
        <button
          type="submit"
          disabled={ticker.length === 0}
          onClick={handleSubmitTicker}
          className={ticker.length === 0 ? "disabled" : ""}
        >
          Submit
        </button>
        <div className="similarStock">
          <label> Show Similar Stocks </label>
          <input
            type="checkbox"
            checked={showSimilarStocks}
            onChange={() => {
              setShowSimilarStocks((prev) => {
                return !prev;
              });
            }}
          />
        </div>

        <button type="submit" onClick={props.handleCreateTable}>
          Create New Table
        </button>
      </div>
      <div className="searchResults">
        {suggestedResults.map((stock) => {
          return (
            <ul
              onClick={() => {
                setTicker(() => stock[0]);
                setSuggestedResults(() => []);
              }}
            >
              Symbol: {stock[0]} Name: {stock[1]}
            </ul>
          );
        })}
      </div>
    </form>
  );
};

export default StockSearch;
