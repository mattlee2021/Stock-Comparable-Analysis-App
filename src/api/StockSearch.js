import { useState } from "react";
import "./StockSearch.scss";
import FetchStocks from "./FetchStocks";
import FetchSimilarStocks from "./FetchSimilarStocks";

const StockApi = (props) => {
  const [showSimilarStocks, setShowSimilarStocks] = useState(false);
  const [ticker, setTicker] = useState("");

  console.log("Ticker StockApi", ticker);

  //API KEY is TD8ZNN64UTNOK6DA

  // Handle empty case by disabling submit if there is no input
  // Now, need to handle case where there is a bad input

  const onSubmitTickerHandler = (event) => {
    event.preventDefault();
    FetchStocks(ticker, props.getStockData);
    if (showSimilarStocks) {
      FetchSimilarStocks(ticker, props.getStockData);
    }
    setTicker("");
  };

  return (
    <form>
      <label> Ticker</label>
      <input
        type="text"
        value={ticker}
        onChange={(event) => {
          setTicker(event.target.value);
        }}
      />
      <button
        type="submit"
        disabled={ticker.length === 0}
        onClick={onSubmitTickerHandler}
        className={ticker.length === 0 ? "disabled" : ""}
      >
        Submit
      </button>
      <label> Show Similar Stocks </label>
      <input
        type="checkbox"
        checked={showSimilarStocks}
        onChange={() => {
          setShowSimilarStocks((prev) => {
            return !prev;
          });
        }}
      />
    </form>
  );
};

export default StockApi;
