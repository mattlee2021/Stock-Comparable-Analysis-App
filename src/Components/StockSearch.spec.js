import TestRenderer from "react-test-renderer";
import StockSearch from "./StockSearch";
import { screen, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import FetchStocks from "../api/FetchStocks";
import "@testing-library/jest-dom";
import FetchSimilarStocks from "../api/FetchSimilarStocks";

jest.mock("../api/FetchStocks");
jest.mock("../api/FetchSimilarStocks");

describe("StockSearch", () => {
  it("Should match the snapshot", () => {
    const testRenderer = TestRenderer.create(<StockSearch />).toJSON();
    expect(testRenderer).toMatchSnapshot();
  });

  it("Should add stock to the stock list", () => {
    const mockGetStockData = jest.fn();
    render(<StockSearch getStockData={mockGetStockData} />);
    userEvent.click(screen.getByRole("checkbox"));
    userEvent.type(screen.getByRole("textbox"), "AAPL");
    userEvent.click(screen.getByRole("button", { name: /submit/i }));
    expect(FetchStocks).toHaveBeenCalledWith("AAPL");
  });

  it("Should call FetchSimilarStocks with the input when the flag is selected", () => {
    const mockGetStockData = jest.fn();
    render(<StockSearch getStockData={mockGetStockData} />);
    userEvent.type(screen.getByRole("textbox"), "AAPL");
    userEvent.click(screen.getByRole("button", { name: /submit/i }));
    expect(FetchSimilarStocks).toHaveBeenCalledWith("AAPL");
  });

  it("Should disable the submit button if there is no inputted text", () => {
    const mockGetStockData = jest.fn();
    render(<StockSearch getStockData={mockGetStockData} />);
    userEvent.type(screen.getByRole("textbox"), "");
    expect(screen.getByRole("button", { name: /submit/i })).toBeDisabled();
  });
});
