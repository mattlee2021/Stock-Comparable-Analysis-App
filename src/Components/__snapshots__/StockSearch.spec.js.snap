// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockSearch Should match the snapshot 1`] = `
<form>
  <div
    className="ticker"
  >
    <label>
      Ticker
    </label>
    <input
      className="searchBar"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="disabled"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
    <div
      className="similarStock"
    >
      <label>
         Show Similar Stocks 
      </label>
      <input
        checked={true}
        onChange={[Function]}
        type="checkbox"
      />
    </div>
    <button
      type="submit"
    >
      Create New Table
    </button>
  </div>
  <div
    className="searchResults"
  />
</form>
`;
